File: ios/TimberCruisingApp/Basemap/TileCache.swift
// Spec §8 Basemap/TileCache. REQ-PRJ-006 pre-download basemap tiles.
//
// A content-addressed on-disk cache: each tile is stored at
//     <cacheRoot>/<providerId>/<z>/<x>/<y>.<ext>
// where `providerId` is a slug derived from the tile URL template so users
// can switch providers without the caches colliding. The cache is plain
// flat files — no database — because offline-first cruises routinely copy
// project folders between devices and a single file tree survives that.
//
// UIKit/MapKit-dependent glue (the MKTileOverlay subclass that reads from
// this cache) lives in TileCache+MapKit.swift in the UI target. Keeping
// the pure-IO layer in the Basemap library target means `swift test` on
// macOS exercises the cache logic directly.

import Foundation
import Geo
#if canImport(CryptoKit)
import CryptoKit
#endif

public final class TileCache: @unchecked Sendable {

    // MARK: - Identity

    public struct Key: Hashable, Sendable {
        public let z: Int
        public let x: Int
        public let y: Int
        public init(z: Int, x: Int, y: Int) {
            self.z = z; self.x = x; self.y = y
        }
    }

    public struct ProviderConfig: Sendable {
        public let urlTemplate: String       // e.g. https://tile.example/{z}/{x}/{y}.png
        public let fileExtension: String     // "png", "jpg"
        public let providerId: String        // slug used as subdirectory

        public init(urlTemplate: String, fileExtension: String, providerId: String) {
            self.urlTemplate = urlTemplate
            self.fileExtension = fileExtension
            self.providerId = providerId
        }

        /// Build a provider id from a URL template if the caller doesn't have
        /// a stable name for it (e.g. user-pasted templates in Settings).
        public static func providerId(forURLTemplate template: String) -> String {
            let lowered = template.lowercased()
            #if canImport(CryptoKit)
            let digest = SHA256.hash(data: Data(lowered.utf8))
            let hex = digest.compactMap { String(format: "%02x", $0) }.joined()
            return "tpl-" + String(hex.prefix(16))
            #else
            return "tpl-" + String(UInt64(bitPattern: Int64(lowered.hashValue)), radix: 36)
            #endif
        }
    }

    // MARK: - State

    public let rootURL: URL
    public let provider: ProviderConfig
    private let fm: FileManager

    public init(rootURL: URL, provider: ProviderConfig, fileManager: FileManager = .default) throws {
        self.rootURL = rootURL
        self.provider = provider
        self.fm = fileManager
        try fm.createDirectory(at: providerDirectory, withIntermediateDirectories: true)
    }

    public var providerDirectory: URL {
        rootURL.appendingPathComponent(provider.providerId, isDirectory: true)
    }

    public func fileURL(for key: Key) -> URL {
        providerDirectory
            .appendingPathComponent("\(key.z)", isDirectory: true)
            .appendingPathComponent("\(key.x)", isDirectory: true)
            .appendingPathComponent("\(key.y).\(provider.fileExtension)")
    }

    // MARK: - Reads / writes

    public func isCached(_ key: Key) -> Bool {
        fm.fileExists(atPath: fileURL(for: key).path)
    }

    public func data(for key: Key) -> Data? {
        guard isCached(key) else { return nil }
        return try? Data(contentsOf: fileURL(for: key))
    }

    public func store(_ data: Data, for key: Key) throws {
        let url = fileURL(for: key)
        try fm.createDirectory(at: url.deletingLastPathComponent(),
                               withIntermediateDirectories: true)
        try data.write(to: url, options: .atomic)
    }

    public func remove(_ key: Key) throws {
        let url = fileURL(for: key)
        if fm.fileExists(atPath: url.path) {
            try fm.removeItem(at: url)
        }
    }

    // MARK: - URL building

    public func resolvedURL(for key: Key) -> URL? {
        let resolved = provider.urlTemplate
            .replacingOccurrences(of: "{z}", with: "\(key.z)")
            .replacingOccurrences(of: "{x}", with: "\(key.x)")
            .replacingOccurrences(of: "{y}", with: "\(key.y)")
        return URL(string: resolved)
    }

    // MARK: - Cache stats

    public struct Stats: Equatable, Sendable {
        public let fileCount: Int
        public let byteCount: Int64
    }

    public func stats() -> Stats {
        var files = 0
        var bytes: Int64 = 0
        if let enumerator = fm.enumerator(at: providerDirectory,
                                          includingPropertiesForKeys: [.fileSizeKey],
                                          options: [.skipsHiddenFiles]) {
            for case let url as URL in enumerator {
                let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey])
                guard values?.isRegularFile == true else { continue }
                files += 1
                bytes += Int64(values?.fileSize ?? 0)
            }
        }
        return Stats(fileCount: files, byteCount: bytes)
    }

    public func clear() throws {
        if fm.fileExists(atPath: providerDirectory.path) {
            try fm.removeItem(at: providerDirectory)
        }
        try fm.createDirectory(at: providerDirectory, withIntermediateDirectories: true)
    }
}

// MARK: - Tile maths

/// Slippy-map tile conversions (OSM / XYZ scheme).
public enum TileMath {

    /// Convert a lat/lon to the tile that contains it at zoom `z`.
    public static func tile(for point: CoordinateConversions.LatLon, zoom z: Int) -> TileCache.Key {
        let n = Double(1 << z)
        let latRad = point.latitude * .pi / 180
        let x = Int(floor((point.longitude + 180) / 360 * n))
        let y = Int(floor((1 - log(tan(latRad) + 1 / cos(latRad)) / .pi) / 2 * n))
        let maxIdx = (1 << z) - 1
        return TileCache.Key(
            z: z,
            x: max(0, min(maxIdx, x)),
            y: max(0, min(maxIdx, y))
        )
    }

    /// North-west corner lat/lon of a tile.
    public static func northwestCorner(of key: TileCache.Key) -> CoordinateConversions.LatLon {
        let n = Double(1 << key.z)
        let lon = Double(key.x) / n * 360 - 180
        let latRad = atan(sinh(.pi * (1 - 2 * Double(key.y) / n)))
        return CoordinateConversions.LatLon(latitude: latRad * 180 / .pi, longitude: lon)
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/basemap/TileCache.kt
// Tile cache — 1:1 port of iOS Basemap/TileCache.swift
// (spec §8, REQ-PRJ-006 pre-download basemap tiles).
//
// A content-addressed on-disk cache: each tile is stored at
//     <cacheRoot>/<providerId>/<z>/<x>/<y>.<ext>
// where `providerId` is a slug derived from the tile URL template so users
// can switch providers without the caches colliding. The cache is plain
// flat files — no database — because offline-first cruises routinely copy
// project folders between devices and a single file tree survives that.
//
// The network + bitmap glue (TileFetcher, the Android analogue of the
// MKTileOverlay subclass) lives in TileFetcher.kt. Keeping the pure-IO
// layer separate mirrors the iOS library/UI target split.

package com.hcjeong.forestix.basemap

import com.hcjeong.forestix.geo.CoordinateConversions
import java.io.File
import java.net.MalformedURLException
import java.net.URL
import java.security.MessageDigest
import java.util.Locale
import kotlin.math.atan
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sinh
import kotlin.math.tan

class TileCache(
    val rootDir: File,
    val provider: ProviderConfig,
) {

    // MARK: - Identity

    data class Key(val z: Int, val x: Int, val y: Int)

    data class ProviderConfig(
        val urlTemplate: String,       // e.g. https://tile.example/{z}/{x}/{y}.png
        val fileExtension: String,     // "png", "jpg"
        val providerId: String,        // slug used as subdirectory
    ) {
        companion object {
            /// Build a provider id from a URL template if the caller doesn't have
            /// a stable name for it (e.g. user-pasted templates in Settings).
            /// Same SHA-256 recipe as the iOS CryptoKit path so both platforms
            /// bucket a given template into the same directory name.
            fun providerId(forURLTemplate: String): String {
                val lowered = forURLTemplate.lowercase()
                val digest = MessageDigest.getInstance("SHA-256")
                    .digest(lowered.toByteArray(Charsets.UTF_8))
                val hex = digest.joinToString("") { String.format(Locale.US, "%02x", it) }
                return "tpl-" + hex.take(16)
            }
        }
    }

    // MARK: - State

    val providerDirectory: File = File(rootDir, provider.providerId)

    init {
        providerDirectory.mkdirs()
    }

    fun fileURL(key: Key): File =
        File(
            File(File(providerDirectory, "${key.z}"), "${key.x}"),
            "${key.y}.${provider.fileExtension}",
        )

    // MARK: - Reads / writes

    fun isCached(key: Key): Boolean = fileURL(key).exists()

    fun data(key: Key): ByteArray? {
        val file = fileURL(key)
        if (!file.exists()) return null
        return try {
            file.readBytes()
        } catch (_: Exception) {
            null
        }
    }

    fun store(data: ByteArray, key: Key) {
        val file = fileURL(key)
        file.parentFile?.mkdirs()
        // Write-then-rename for the atomicity iOS gets from `.atomic`.
        val tmp = File(file.parentFile, "${file.name}.tmp")
        tmp.writeBytes(data)
        if (!tmp.renameTo(file)) {
            file.writeBytes(data)
            tmp.delete()
        }
    }

    fun remove(key: Key) {
        val file = fileURL(key)
        if (file.exists()) file.delete()
    }

    // MARK: - URL building

    fun resolvedURL(key: Key): URL? {
        val resolved = provider.urlTemplate
            .replace("{z}", "${key.z}")
            .replace("{x}", "${key.x}")
            .replace("{y}", "${key.y}")
        return try {
            URL(resolved)
        } catch (_: MalformedURLException) {
            null
        }
    }

    // MARK: - Cache stats

    data class Stats(val fileCount: Int, val byteCount: Long)

    fun stats(): Stats {
        var files = 0
        var bytes = 0L
        providerDirectory.walkTopDown().forEach { f ->
            if (f.isFile && !f.name.startsWith(".")) {
                files += 1
                bytes += f.length()
            }
        }
        return Stats(fileCount = files, byteCount = bytes)
    }

    fun clear() {
        if (providerDirectory.exists()) {
            providerDirectory.deleteRecursively()
        }
        providerDirectory.mkdirs()
    }
}

// MARK: - Tile maths

/// Slippy-map tile conversions (OSM / XYZ scheme).
object TileMath {

    /// Convert a lat/lon to the tile that contains it at zoom `z`.
    fun tile(point: CoordinateConversions.LatLon, zoom: Int): TileCache.Key {
        val n = (1 shl zoom).toDouble()
        val latRad = point.latitude * Math.PI / 180
        val x = floor((point.longitude + 180) / 360 * n).toInt()
        val y = floor((1 - ln(tan(latRad) + 1 / cos(latRad)) / Math.PI) / 2 * n).toInt()
        val maxIdx = (1 shl zoom) - 1
        return TileCache.Key(
            z = zoom,
            x = max(0, min(maxIdx, x)),
            y = max(0, min(maxIdx, y)),
        )
    }

    /// North-west corner lat/lon of a tile.
    fun northwestCorner(key: TileCache.Key): CoordinateConversions.LatLon {
        val n = (1 shl key.z).toDouble()
        val lon = key.x / n * 360 - 180
        val latRad = atan(sinh(Math.PI * (1 - 2 * key.y / n)))
        return CoordinateConversions.LatLon(latitude = latRad * 180 / Math.PI, longitude = lon)
    }
}
